app/src/main/java/at/bitfire/icsdroid/ui/CredentialsComposable.kt

```diff
@@ -35,12 +35,18 @@ import at.bitfire.icsdroid.R
 @Composable
 fun LoginCredentialsComposable(
     requiresAuth: Boolean,
-    username: String?,
-    password: String?,
+    username: String? = null,
+    password: String? = null,
     onRequiresAuthChange: (Boolean) -> Unit,
     onUsernameChange: (String) -> Unit,
     onPasswordChange: (String) -> Unit
 ) {
+    val usernameError = if (username?.isBlank() == true)
```

Comment:
We can just use [`isNullOrBlank`](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.text/is-null-or-blank.html)

Reply:
No, we can not, actually. Otherwise the text error field will be marked erroneous, right away, without the user having entered anything yet. 

I think it should only become an error if the value is empty or blank (`""`), and not if it is `null` :)

Reply:
@ArnyminerZ Do you maybe know some cleaner way to make it work like that?

Reply:
I'll take a look

Reply:
Look at what I've added in 5cba9ff. I think it works pretty fine.

When the switch is enabled (aka `requiresAuth` is true), and the fields are empty, they are shown in red, and the error message is shown below (instead than on the label, which was a bit confusing):

![errors-1](https://github.com/ArnyminerZ/icsx5/assets/12086466/2137f76b-4865-441b-8109-f4bd8cb715af)

And at the moment you introduce something, it just disappears:

![errors-2](https://github.com/ArnyminerZ/icsx5/assets/12086466/de2e3d67-1fd8-49af-876b-585aaa17e160)

I feel that it's more intuitive and illustrating.

Reply:
Now that I think about it, I think we even can simplify
```kotlin
val usernameError = if (requiresAuth && username.isNullOrBlank())
        stringResource(R.string.edit_calendar_need_username)
    else null
```
to
```kotlin
val usernameError = if (username.isNullOrBlank())
        stringResource(R.string.edit_calendar_need_username)
    else null
```
(removing `requiresAuth`) since the error is never shown if it's true.

Reply:
> Look at what I've added in [5cba9ff](https://github.com/ArnyminerZ/icsx5/commit/5cba9fff34812b07b6cc459ad3df6d874be04c42). I think it works pretty fine.

I don't like thes changes in https://github.com/ArnyminerZ/icsx5/commit/5cba9fff34812b07b6cc459ad3df6d874be04c42 at all. Here is why:

> When the switch is enabled (aka `requiresAuth` is true), and the fields are empty, they are shown in red,

Showing the error right away:

 Again, as I said before: If we show the error before the user has even had the chance to give input, it feels like punishing the user without them having done anything wrong. Here is an article on [premature input validation](https://www.smashingmagazine.com/2022/08/error-messages-ux-design/#in-forms-display-error-messages-above-input) and why it is bad for UX.

> ... and the error message is shown below (instead than on the label, which was a bit confusing):

Showing the error below the input field. 
  1. Testing this on my emulator - the focus always makes the textbox scroll right above the keyboard, hereby covering up the error message. I can not see the error, as it gets completely covered up by the keyboard. Here is another UX article mentioning that it's better to [put the error message above the text field](https://www.smashingmagazine.com/2022/08/error-messages-ux-design/#in-forms-display-error-messages-above-input) because of other (and the same) reasons.
  2. Not having the error message in the label, is confusing for which textfield it belongs to. We could put it above the input field and use extra spacing, between the text fields, but when using the label it is completely clear which text field the error belongs to. The way I had it, the error message repeats the label name (Label: "Password", Error: "Password required". Why do you find it confusing to be in the label?


> And at the moment you introduce something, it just disappears:

Yes, it works this way in both our variants. 

> I feel that it's more intuitive and illustrating.

Try to see the very short video here, I find this to be much more intuitive an illustrating. :)
https://baymard.com/blog/inline-form-validation#3-use-positive-inline-validation-for-all-fields

Reply:
I just checked the [material guidelins on text boxes and errors](https://m2.material.io/components/text-fields#filled-text-field) too, and I can see they want the error message to go below as well. I don't think that's good, because of above mentioned reasons. 

They also want the error message to replace an existing helper text which we don't have either, so adding an error message always makes the view move.

And [this](https://stackoverflow.com/questions/31047449/error-text-in-textinputlayout-is-covered-by-keyboard#answer-58326757) is ridiculous. It seems to be a problem since 2019 which Google does not want to address https://issuetracker.google.com/issues/37051832

Reply:
> If we show the error before the user has even had the chance to give input

It doesn't. At first no error is displayed. The error just appears once you have introduced something into the field. Actually what happens is that the error doesn't show up if `username` or `password` is null, at the moment you introduce something, even if you clear the field, the value would be `""` and not `null`, and that's why then it shows up.

> the focus always makes the textbox scroll right above the keyboard

That can be fixed without much issues. A bit of boilerplate code, but not too much.

> Why do you find it confusing to be in the label?

Because the color of the label is always the "idle" one, it would be more illustrative if it was also red, which we can absolutely do.

Maybe instead of using the label, we can use the placeholder for the error. The error only shows when the field is empty, so that would work out. It would be inside the field and not cause any confusion. I still don't know what the color behavior would be.

> They also want the error message to replace an existing helper text which we don't have either, so adding an error message always makes the view move.

So what do we do? Ignore the guidelines and go for a "more correct" and "UX-friendly" approach? Or just make what Google wants?

Reply:
Maybe we can also label the fields with a little red `*` which is usually associated with required fields.

Reply:
The placeholder approach looks quite good to me:

![username](https://github.com/ArnyminerZ/icsx5/assets/12086466/93fa97ee-30be-4cf7-ba22-d2b91813ee92)

Reply:
> > If we show the error before the user has even had the chance to give input
> 
> ~It doesn't~ (_edit: it shouldn't_). At first no error is displayed. The error just appears once you have introduced something into the field. Actually what happens is that the error doesn't show up if `username` or `password` is null, at the moment you introduce something, even if you clear the field, the value would be `""` and not `null`, and that's why then it shows up.

Exactly, that is why I did not use `isNullOrBlank()`. In https://github.com/ArnyminerZ/icsx5/commit/5cba9fff34812b07b6cc459ad3df6d874be04c42 , however, that is not the case. The error shows up even if it is null.  
So I think we should go back to using `val usernameError = if (username?.isBlank() == true)`. :)

> So what do we do? Ignore the guidelines and go for a "more correct" and "UX-friendly" approach? Or just make what Google wants?

I think we should always strive for the "most correct" and UX friendly way (most often this alignes with what Google or the Material Guidelines say anyways), while keeping a balance of not moving too far from sensible defaults (think less code and less maintenance if something changes). 

I don't think it's good to blindly follow all of googles advices - especially not if it means more work now and in the future ... In this case I believe we should deviate from the standard and find the balance between good UX and a reasonable amount of code (less is better). 

With this in mind:

> > the focus always makes the textbox scroll right above the keyboard
> 
> That can be fixed without much issues. A bit of boilerplate code, but not too much.

I would not like to add more code than necessary. Especially not for such a trivial thing.

> Maybe instead of using the label, we can use the placeholder for the error. The error only shows when the field is empty, so that would work out. It would be inside the field and not cause any confusion. I still don't know what the color behavior would be.

Good idea, but that would not be consistent with how we might have to show other errors on text fields (also thinking about DAVx5 here already). So, other input validation messages, like "Invalid email" or "Password too short" should show in the same spot.  We should stay consistent throughout the whole app.

> > Why do you find it confusing to be in the label?
> 
> Because the color of the label is always the "idle" one, it would be more illustrative if it was also red, which we can absolutely do.

Really? For me the color of the label turns red as well :thinking: But yes, I think we should make it red too, if it does not work on your device.

![image](https://github.com/ArnyminerZ/icsx5/assets/8227414/56f492e1-9261-4aaf-9070-8e5d76ce0117)


Just so we are clear: If Google had provided an easier and integrated way of showing errors, I would have preferred the google way. :)

Reply:
Ah, you meant while the label does not turn red while it is inside the text field ...

![image](https://github.com/ArnyminerZ/icsx5/assets/8227414/a8cb2cd3-80b4-430a-9ec8-0565deacc1b2)

I don't think it to be that confusing, but yeah, could make it red as well

Reply:
Just had a call with Ricki and mentioned this. 
Since we are going to migrate to M3 anyways, I think we can just this as it is for now.

[M3 Text fields](https://m3.material.io/components/text-fields/overview) have [proper support](https://developer.android.com/reference/kotlin/androidx/compose/material3/package-summary#textfield) for `supportingText`. Which we should use then :)

Reply:
> > If we show the error before the user has even had the chance to give input
> 
> ~It doesn't~ (_edit: it shouldn't_). At first no error is displayed. The error just appears once you have introduced something into the field. Actually what happens is that the error doesn't show up if `username` or `password` is null, at the moment you introduce something, even if you clear the field, the value would be `""` and not `null`, and that's why then it shows up.

Since we seem to have been talking about the same thing I just changed the `isNullOrBlank()` checks back to `(username?.isBlank() == true)` so it works again. 

Should be ready for merge now :)